drivers: sensors: adt7420: Add driver for ADT7420 Temperature Sensor

mhennerich opened the pull request:
This patch adds support for the Analog Devices ADT7420 High-Accuracy
16-bit Digital I2C Temperature Sensors. Optionally sensor threshold
events are supported.

Signed-off-by: Michael Hennerich <michael.hennerich@analog.com>

MaureenHelm commented on drivers/sensor/adt7420/adt7420.c:
```diff
@@ -0,0 +1,202 @@
+		/* Fallthrough */
+	case SENSOR_ATTR_LOWER_THRESH:
+		if (!reg)
+			reg = ADT7420_REG_T_LOW_MSB;
```
Please wrap in braces

mhennerich replied:
Please bear with me, that's my first Zephyr patch. I thought that we follow Linux kernel coding style.
Braces around single line code blocks seems to be an exception...

MaureenHelm replied:
Yes, this is an exception to the Linux coding style.

MaureenHelm commented on drivers/sensor/adt7420/adt7420.c:
```diff
@@ -0,0 +1,202 @@
+			reg = ADT7420_REG_T_LOW_MSB;
+
+		value = (s64_t)val->val1 * 1000000 + val->val2;
+		value = (value / ADT7420_TEMP_SCALE) << 1;
```
Is it possible this result could be out of range before you cast it to s16_t and pass to `adt7420_temp_reg_write`?

mhennerich replied:
No - not if you pass valid threshold values.

MaureenHelm replied:
Then you should add a check to ensure the threshold values are valid

MaureenHelm commented on drivers/sensor/adt7420/adt7420.c:
```diff
@@ -0,0 +1,202 @@
+	ret = i2c_reg_read_byte(drv_data->i2c, drv_data->i2c_addr,
+				ADT7420_REG_ID, &value);
+	if (ret)
+		return ret;
```
Please add braces after all the if statements.
http://docs.zephyrproject.org/contribute/contribute_guidelines.html#coding-style

mhennerich replied:
sure

MaureenHelm commented on drivers/sensor/adt7420/adt7420.c:
```diff
@@ -0,0 +1,202 @@
+	return 0;
+}
+
+int adt7420_init(struct device *dev)
```
This can be static like the other functions.

mhennerich replied:
good catch

MaureenHelm commented on drivers/sensor/adt7420/adt7420.c:
```diff
@@ -0,0 +1,202 @@
+		return -EINVAL;
+	}
+
+	drv_data->i2c_addr = CONFIG_ADT7420_I2C_ADDR;
```
Move the i2c addr to a const config structure and remove this line. You can also add the i2c_dev name to the const config structure instead of passing `CONFIG_ADT7420_I2C_MASTER_DEV_NAME` directly to `device_get_binding`

mhennerich replied:
Thanks. Makes sense to keep the config stuff in one place.

MaureenHelm commented on drivers/sensor/adt7420/adt7420_trigger.c:
```diff
@@ -0,0 +1,131 @@
+#include <sensor.h>
+#include "adt7420.h"
+
+extern struct adt7420_data adt7420_driver;
```
Why is this here?

mhennerich replied:
not needed at all

MaureenHelm commented on drivers/sensor/adt7420/adt7420_trigger.c:
```diff
@@ -0,0 +1,131 @@
+{
+	struct adt7420_data *drv_data = dev->driver_data;
+
+	drv_data->gpio = device_get_binding(CONFIG_ADT7420_GPIO_DEV_NAME);
```
`CONFIG_ADT7420_GPIO_DEV_NAME` and `CONFIG_ADT7420_GPIO_PIN_NUM` can also be stored in the const config structure

mhennerich replied:
will do.

MaureenHelm reviewed (changes requested):
Please add a sample or update tests/drivers/build_all/sensors_a_m.conf to include this driver. This will ensure the driver gets built in CI. Are there any in-tree boards that have this sensor?

mhennerich commented on drivers/sensor/adt7420/Kconfig:
```diff
@@ -0,0 +1,117 @@
+	  Enable the driver for Analog Devices ADT7420 High-Accuracy
+	  16-bit Digital I2C Temperature Sensors.
+
+config ADT7420_NAME
```
Hi Maureen,

Thanks for your feedback!

In theory device tree is used to describe bus hierarchies and devices that can appear under a bus. 
Basically it describes the physical devices in a system, which cannot be automatically probed.

I totally agree that the parent bus and slave addresses, etc. should come from the device tree.
Practically this would come from the parent device which will instantiate all its devices, via a device specific structure.
In fact that’s the only way to allow instantiating the same driver multiple times…
Which I have no idea, how that can be done in Zephyr today.
 
This being said,  Zephyr device tree usage is very limited. The user must provide a fixup file which maps
device tree properties to CONFIG options.

So my question is what do you mean by “should come from the device tree” ?
Isn’t what’s implemented in this driver exactly the same what all other drivers/sensor do?

MaureenHelm replied:
It's been a long process converting drivers over to device tree, so you will find that some drivers are further along than others. For new drivers, however, we do want to enforce getting hardware information from device tree from the start. Please have a look at `drivers/sensors/fxos8700` and `boards/arm/frdm_k64f/frdm_k64f.dts` for an example of what we're looking for.

mhennerich replied:
Oh I see - you asking me to guard these config options with "if !HAS_DTS_I2C_DEVICE" !?

MaureenHelm replied:
Yes, and to add a binding in `dts/bindings/sensor/`

mhennerich commented on drivers/sensor/adt7420/Kconfig:
```diff
@@ -0,0 +1,117 @@
+	string
+	prompt "Driver name"
+	default "ADT7420"
+	depends on ADT7420
```
Will do

mhennerich commented on drivers/sensor/adt7420/Kconfig:
```diff
@@ -0,0 +1,117 @@
+	int "Critical overtemperature in °C"
+	depends on ADT7420
+	range -40 150
+	default 147
```
Yes 147 °C is the critical over temperature reset default of the ADT7420. So if no one changes this config, it'll use the device default.

MaureenHelm replied:
Ok, cool. Will you please note this in the help?

mhennerich commented on drivers/sensor/adt7420/Kconfig:
```diff
@@ -0,0 +1,117 @@
+	bool
+	depends on ADT7420
+
+config ADT7420_GPIO_DEV_NAME
```
Please see previous question

mhennerich commented:
currently there are no in-tree boards with this sensor.
Shall I merge my samples/sensor/adt7420 example into this PR or shall I open a separate one?

Thanks,
Michael

MaureenHelm commented:
> Shall I merge my samples/sensor/adt7420 example into this PR or shall I open a separate one?

Please add it in another commit in this PR.

mhennerich committed:
drivers: sensors: adt7420: Add driver for ADT7420 Temperature Sensor

This patch adds support for the Analog Devices ADT7420 High-Accuracy
16-bit Digital I2C Temperature Sensors. Optionally sensor threshold
events are supported.

Signed-off-by: Michael Hennerich <michael.hennerich@analog.com>

mhennerich commented:
Hi,

I think I addressed all the feedback and issues.
However CI fails now on some platforms which I guess using HAS_DTS_I2C_DEVICE.

/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/sensor/adt7420/adt7420.c:214:14: error: ‘CONFIG_ADT7420_I2C_MASTER_DEV_NAME’ undeclared here (not in a function)

Wondering how shall we avoid these expected failing CI tests?

-Michael

mhennerich commented:
Ok never mind - I fixed (and pushed) the build issue on boards using CONFIG_HAS_DTS_I2C_DEVICE, by providing  a dummy dts.fixup in case some of the required CONFIG_ADT7240_ defines are not present in the device tree.

MaureenHelm commented on drivers/sensor/adt7420/Kconfig:
```diff
@@ -0,0 +1,123 @@
+config ADT7420_TRIGGER
+	bool
+
+if !HAS_DTS_I2C_DEVICE
```
This should be `if !HAS_DTS_GPIO_DEVICE`

mhennerich replied:
right

MaureenHelm commented on samples/sensor/adt7420/src/main.c:
```diff
@@ -0,0 +1,97 @@
+
+void main(void)
+{
+	struct device *dev = device_get_binding("ADT7420");
```
Use `CONFIG_ADT7420_NAME` instead of the hardcoded string

MaureenHelm commented on samples/sensor/adt7420/src/main.c:
```diff
@@ -0,0 +1,97 @@
+		printf("temperature %.6f C\n",
+		       sensor_value_to_double(&temp_val));
+
+		k_sleep(1000);
```
You probably don't want this if trigger is enabled

mhennerich replied:
Yep

MaureenHelm commented on samples/sensor/adt7420/README.txt:
```diff
@@ -0,0 +1,5 @@
+
+Description:
+
+A example using the ADT7420 high accuracy digital temperature sensor.
```
I expect @dbkinder will ask for more details, like how to build the sample and what is the expected output.

mhennerich replied:
Just looking at the other examples - I'll do the whole thing including reference, wiring building and running. Thanks for your review. I'll push the fixes right now.

dbkinder commented on samples/sensor/adt7420/README.rst:
```diff
@@ -0,0 +1,55 @@
+ADT7420: High accuracy digital I2C temperature sensor
+#####################################################
+
+Desciption
```
``Description``

mhennerich replied:
Good catch

dbkinder commented on samples/sensor/adt7420/README.rst:
```diff
@@ -0,0 +1,55 @@
+
+This sample application periodically (1Hz) measures the ambient temperature
+in degrees Celsius. The result is written to the console.
+Optionally it also shows how to use the upper and lower threshold triggers.
```
Add a comma, ``Optionally, it``

mhennerich replied:
ok

dbkinder commented on samples/sensor/adt7420/README.rst:
```diff
@@ -0,0 +1,55 @@
+
+This sample uses the ADT7420 sensor controlled using the I2C interface.
+Connect Supply: **VDD**, **GND** and Interface: **SDA**, **SCL**
+optionally connect the **INT** to a interrupt capabale GPIO.
```
add ``and optionally``

``capable``

mhennerich replied:
Will do. Thanks for the Review!
I already fixed and pushed them.

dbkinder reviewed (dismissed):
pretty close on documentation... a few misspellings and punctuation fixes.  Thanks!

mhennerich committed:
samples: sensor: adt7420: Add ADT7420 sample application

This simple application periodically prints the ambient temperature.
Optional support for threshold triggers is provided.

Signed-off-by: Michael Hennerich <michael.hennerich@analog.com>
